src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -2614,6 +2614,28 @@ public static T MappedAs<T>(this T parameter, IType type)
 			throw new InvalidOperationException("The method should be used inside Linq to indicate a type of a parameter");
 		}
 
+		public static IQueryable<T> AsNoTracking<T>(this IQueryable<T> query)
+		{
+			if (query is NhQueryable<T> nhQuery)
+			{
+				var oldProvider = nhQuery.Provider as DefaultQueryProvider;
+
+				if (oldProvider.Session is IStatelessSession)
+				{
+					return query;
+				}
+
+				var statelessSession = (oldProvider.Session as SessionImpl)
+					.StatelessSessionWithOptions()
+					.Connection(oldProvider.Session.Connection)
+					.OpenStatelessSession() as ISessionImplementor;
```

Comment:
I believe all castings can be dropped in this code using something like this:
```suggestion
				var statelessSession = oldProvider.Session
					.StatelessSessionWithOptions()
					.Connection()
					.OpenStatelessSession()
					.GetSessionImplementation()
```

And it looks like it won't work properly for `Future` queries  (when both `AsNoTracking` and common queries are present in batch).

Reply:
It does work with ToFuture, included a unit test. As for your code, it does not compile.

Reply:
> It does work with ToFuture,

Your test does not cover scenario I described - when both AsNoTracking and common queries are present in batch. Multiple future queries should be executed as one db roundtrip. I believe It won't happen with you changes (simply because `Future` queries are batched in session batcher - and your implementation just creates new session for each no tracking query)

> it does not compile. 

Missing semicolon or what? Yes I didn't compile it and give it as example -  I thought you would able to figure out what's missing.

Reply:
Ok I see. `oldProvider.Session` returns implementor so one cast is still required: `(ISession) oldProvider.Session`

Reply:
Your remark about mixing AsNoTracking and common queries might be true, haven't tested it, but I don't see that as a problem. I see people using AsNoTracking when all they have is an ISession and they want, for example, to populate a list on a web page, that is, when they don't need the entities to be tracked. Anyway, that could always be mentioned in the documentation.

Reply:
Well, every extension method is a shortcut, so, in that sense, yes. But if you actually look at the code you can see that it's a bit more complex than that.